### PR TITLE
Add Conditions for ActiveGate and CodeModules to the Istio Reconciler

### PR DESCRIPTION
**Pull request by wepudt**
## Description

[K8S-9578](https://dt-rnd.atlassian.net/browse/K8S-9578?atlOrigin=eyJpIjoiMzk3NmZmOGU3MDQ1NGIwYjg3MjA3MzlhZGE0MTUyMzUiLCJwIjoiaiJ9)

Have conditions for the istio reconciler, for the ActiveGate and CodeModule connection-info ServiceEntry creation logic.
Have “Reasons“ for the following scenarios:

* ServiceEntries and VirualServices created/updated

* Error due to kubernetes API problems

## How can this be tested?
 
* Run the unit-tests.
* Deploy Istio and a Dynakube with Istio enabled. Describe the DynaKube and look for the newly added conditions starting with `IstioServiceConfigurationFor*`.

### CODE REVIEW

File: pkg/controllers/dynakube/istio/conditions.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package istio
+
+import (
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func getConditionTypeName(component string) string {
+	return "IstioServiceConfigurationFor" + component
```
Comment (0sewa0) [inline pkg/controllers/dynakube/istio/conditions.go:9]: this feels a bit too verbose, my suggestions for alternatives:
- `<component>IstioConfig`
- `IstioFor<component>`

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 2]
```diff
@@ -62,11 +63,21 @@ func (r *reconciler) ReconcileCodeModuleCommunicationHosts(ctx context.Context,
 
 	oneAgentCommunicationHosts := oaconnectioninfo.GetCommunicationHosts(dynakube)
 
+	const conditionComponent = "CodeModule"
```
Comment (0sewa0) [inline pkg/controllers/dynakube/istio/reconciler.go:66]: couldn't we reuse the `OneAgentComponent` const here? 
I know that technically `CodeModule` is more accurate

but maybe we could change the `OneAgentComponent` to be more accurate
- Only do this if it does not break any naming

otherwise we could still just reuse `OneAgentComponent` and give it the old `strings.ToTitle()` to make it more "presentable" 😅

Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go:66]: I actually did that. Turns out `toTitle(..)` uppercases all letters, not just the initial one. :(

Reusing the existing constants would either mean to 
1. have the component part all lowercase, 
2. or all uppercase (using `toTitle(...)`)
3. or do some string-voodoo to just uppercase the initial rune.
4. 
possibilities 1 and 2 seemed bad in respect to readability, 3rd one seemed a bit over-engineered so the simplest solution I could think of was to use a constant... (changing the existing one seemed the worse idea as it would break exisiting behaviour)

File: pkg/controllers/dynakube/istio/conditions.go (hunk 1 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/conditions.go]: fixed in #e131fa0d

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 2 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/istio/reconciler.go:66]: Did a bit of research, and yes the `ToTitle()` func is confusing AF.
There is a `.Title()` func that does kinda what I want. (makes only the first char in a word upper case) 
But even then, we would end up with the `Codemodule` or `Oneagent`

 so lets leave it as is, and do some refactoring incase in the future it will confuse us

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 3]
```diff
@@ -83,11 +94,21 @@ func (r *reconciler) ReconcileActiveGateCommunicationHosts(ctx context.Context,
...
 	if err != nil {
+		setServiceEntryFailedConditionForComponent(dynakube.Conditions(), conditionComponent, err)
+
 		return err
 	}
 
+	if len(activeGateEndpoints) > 0 {
+		setServiceEntryUpdatedConditionForComponent(dynakube.Conditions(), conditionComponent)
+	} else {
+		meta.RemoveStatusCondition(dynakube.Conditions(), getConditionTypeName(conditionComponent))
```
Comment (0sewa0) [inline pkg/controllers/dynakube/istio/reconciler.go:109]: one of the Acceptance Criteria in the linked ticket:
- `condition and istio resources are cleaned up, when not necessary`

you do a clean up here for the condition (and in line 107, the istio stuff is also cleaned up)

but in line 91, we do not clean up, if the activeGate was "turned off", and we should

(also the log line in line 88 is incorrect, but that is a mistake done by someone else 😅 )

the clean-up should also be double-checked in the `ReconcileCodeModuleCommunicationHosts` func, but maybe it just works 🤔

Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go:111]: fixed in 2d824c6d

I think that this is no issue for the CodeModules as the conditions are 
* set to failed if there were errors
* set to updated if there are CommunicationHosts where the ServiceEntries were updated
* removed if there are none

Comment (0sewa0) [reply pkg/controllers/dynakube/istio/reconciler.go:111]: the `if !dynakube.NeedsActiveGate` check ([line 93-97](https://github.com/Dynatrace/dynatrace-operator/pull/3198/files#diff-6dcf84623910535feca0c7dc62e7a35bd7706001d348bc3c88ba5f69893b100cL80-L82)) will stop the cleanup of the istio resources we created for the ActiveGate, as the clean up happens deeper in the `r.reconcileCommunicationHostsForComponent` :
- [here](https://github.com/Dynatrace/dynatrace-operator/blob/e93ae0ad8f1a0552cfce5ff7ee5818809fe3e6d8/pkg/controllers/dynakube/istio/reconciler.go#L150-L157) and [here](https://github.com/Dynatrace/dynatrace-operator/blob/e93ae0ad8f1a0552cfce5ff7ee5818809fe3e6d8/pkg/controllers/dynakube/istio/reconciler.go#L184-L194)

we don't have a similar check for the CodeModules, so those get cleaned up (probably, only looked at code, didn't run tests)

the most "efficient" way to handle the creation/update/delete with the help of the condition: (we already do this for 1 or 2 secrets)
1. `if needed && conditionNotSet` => Create resources and set condition
2. `if needed && conditionPresent && conditionUpToDate` => do nothing
3. `if needed && conditionPresent && conditionNOTUpToDate` => update resources and condtion
4. `if notNeeded && conditionPresent` => cleanup resources and condition
5. `if notNeeded && conditionNotPresent` => do nothing

(it looks something like [this](https://github.com/Dynatrace/dynatrace-operator/blob/290c4dac9064a4c97be150f373c8e39c7dcd6aed/pkg/controllers/dynakube/connectioninfo/activegate/reconciler.go#L43-L69), its not just a bunch of `ifs`, more of a flow)

Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go:123]: fixed in ecac0abd

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 4]
```diff
@@ -62,35 +67,89 @@ func (r *reconciler) ReconcileCodeModuleCommunicationHosts(ctx context.Context,
...
+	// try to clean up all entries even if one fails
+	if err1 != nil {
+		return err1
+	}
+
+	if err2 != nil {
+		return err2
+	}
+
 	return nil
```
Comment (0sewa0) [inline pkg/controllers/dynakube/istio/reconciler.go:144]:
```suggestion
	return errors.Join(err1, err2)
```

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 5]
```diff
@@ -62,35 +67,89 @@ func (r *reconciler) ReconcileCodeModuleCommunicationHosts(ctx context.Context,
...
+		if isIstioConfigured(dynakube, conditionComponent) {
+			log.Info("ActiveGate disabled, cleaning up")
+
+			return r.CleanupIstio(ctx, dynakube, conditionComponent)
+		} else {
+			return nil
+		}
+	}
+
+	if !conditions.IsOutdated(r.timeProvider, dynakube, getConditionTypeName(conditionComponent)) {
```
Comment (0sewa0) [inline pkg/controllers/dynakube/istio/reconciler.go:107]: I like this, could we do this for the ` ReconcileCodeModuleCommunicationHosts`?

you should just need to change the `if` to `!dk.NeedAppInjection()`

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 4 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: fixed in f30e0998

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 6]
```diff
@@ -2,14 +2,18 @@ package istio
...
 
 	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta1/dynakube"
 	dtclient "github.com/Dynatrace/dynatrace-operator/pkg/clients/dynatrace"
 	"github.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube/connectioninfo/activegate"
 	"github.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube/connectioninfo/oneagent"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/conditions"
 	"github.com/Dynatrace/dynatrace-operator/pkg/util/kubeobjects/labels"
-	"github.com/pkg/errors"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/timeprovider"
+	errorshelper "github.com/pkg/errors"
```
Comment (0sewa0) [inline pkg/controllers/dynakube/istio/reconciler.go:15]: pretty sure you can just remove this and use `"errors"` instead

Comment (0sewa0) [reply pkg/controllers/dynakube/istio/reconciler.go:15]: the `WithMessage` func is missing in `"errors"` 😭

Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go:15]: that's why I just left it in... we decided to postpone the change to a different story, but after doing some fast research I found, that there are a bunch of libs, each with advantages and disadvantages. The one we use is pretty simple so I would just keep it for now...

Comment (0sewa0) [inline pkg/controllers/dynakube/istio/reconciler.go:15]: just to be consistent:

we tend to alias the `errors` to `goerrors`
and leave `github.com/pkg/errors` as `errors`

https://github.com/Dynatrace/dynatrace-operator/blob/7d61cb9fed001f98c64157b193f4bc25eca742cf/pkg/controllers/dynakube/controller.go#L5

Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: fixed in cad3eb41

**Comment by wepudt** [comment]
> pls rebase changes to main to resolve the conflict
> 
> tried it out, and the cleanup does not work because `ReconcileCodeModuleCommunicationHosts` and `ReconcileActiveGateCommunicationHosts` are called in block that checks `if featuresIsNeeded`, which is not necessary.
> 
> * https://github.com/Dynatrace/dynatrace-operator/blob/7d61cb9fed001f98c64157b193f4bc25eca742cf/pkg/controllers/dynakube/activegate/reconciler.go
> * https://github.com/Dynatrace/dynatrace-operator/blob/7d61cb9fed001f98c64157b193f4bc25eca742cf/pkg/controllers/dynakube/injection/reconciler.go

fixed in cad3eb41

File: pkg/controllers/dynakube/injection/reconciler.go [hunk 7]
```diff
@@ -78,6 +78,14 @@ func (r *reconciler) Reconcile(ctx context.Context) error {
...
 	if !r.dynakube.NeedAppInjection() {
+		if r.istioReconciler != nil {
+			err = r.istioReconciler.ReconcileCodeModuleCommunicationHosts(ctx, r.dynakube)
+
+			if err != nil {
+				log.Info("Error reconciling Istio codemodules", "error", err)
+			}
+		}
+
 		return r.removeAppInjection(ctx)
```
Comment (0sewa0) [inline pkg/controllers/dynakube/injection/reconciler.go:89]: why are you only reconciling istio when no app injection is needed? -> this will cause to only clean up things, but never create them

you should try to reconcile in every case so we create/update when we need to or cleanup

Comment (wepudt) [reply pkg/controllers/dynakube/injection/reconciler.go:89]: actually this is due to the way we do reconciliation.

I wanted to have minimal invasive changes. Before we had a short to just check if AppInjection is disabled here.

The actual reconciliation is done way down in `setupOneAgentInjection` but that never got called due to the if further up...

Comment (0sewa0) [reply pkg/controllers/dynakube/injection/reconciler.go:89]: moving it from `setupOneAgentInjection` to here without the is IMO less invasive than having it 2 times in different places.

Comment (luhi-DT) [reply pkg/controllers/dynakube/injection/reconciler.go:89]: Also its checked twice now, here and in the function you check for `!r.dynakube.NeedAppInjection()`

Comment (wepudt) [reply pkg/controllers/dynakube/injection/reconciler.go]: fixed in 94a715a4

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 8]
```diff
@@ -157,40 +221,62 @@ func (r *reconciler) reconcileIPServiceEntry(ctx context.Context, ipHosts []dtcl
...
+	return nil
+}
+
 func (r *reconciler) reconcileFQDNServiceEntry(ctx context.Context, fqdnHosts []dtclient.CommunicationHost, component string) error {
 	owner := r.client.Owner
 	entryName := BuildNameForFQDNServiceEntry(owner.GetName(), component)
 
 	if len(fqdnHosts) != 0 {
-		meta := buildObjectMeta(
+		objectMeta := buildObjectMeta(
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:240]: out of curiosity, why do we need this change here? it seems unrelated to this PR.

File: pkg/controllers/dynakube/injection/reconciler.go [hunk 9]
```diff
@@ -77,6 +77,15 @@ func (r *reconciler) Reconcile(ctx context.Context) error {
 		return err
 	}
 
+	// do istio reconciliation for CodeModules here to enable cleanup of conditions
+	if r.istioReconciler != nil {
+		err = r.istioReconciler.ReconcileCodeModuleCommunicationHosts(ctx, r.dynakube)
+
+		if err != nil {
+			log.Info("Error reconciling Istio configuration for CodeModules", "error", err)
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/injection/reconciler.go:85]:
```suggestion
			log.Info("error reconciling istio configuration for codeModules", "error", err)
```

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 10]
```diff
@@ -54,43 +60,101 @@ func (r *reconciler) ReconcileAPIUrl(ctx context.Context, dynakube *dynatracev1b
...
+			log.Info("ActiveGate disabled, cleaning up")
+
+			return r.CleanupIstio(ctx, dynakube, conditionComponent, ActiveGateComponent)
+		} else {
+			return nil
+		}
+	}
+
+	if !conditions.IsOutdated(r.timeProvider, dynakube, getConditionTypeName(conditionComponent)) {
+		log.Info("Condition still within time threshold...skipping further reconciliation")
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:119]:
```suggestion
		log.Info("condition still within time threshold...skipping further reconciliation")
```

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 11]
```diff
@@ -135,20 +199,20 @@ func (r *reconciler) reconcileIPServiceEntry(ctx context.Context, ipHosts []dtcl
 	entryName := BuildNameForIPServiceEntry(owner.GetName(), component)
 
 	if len(ipHosts) != 0 {
-		meta := buildObjectMeta(
+		objectMeta := buildObjectMeta(
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:202]: same as above.

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 8 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go:240]: true, but I got the hint `Variable 'meta' collides with imported package name` so I thought it might improve readability...

Comment (andriisoldatenko) [reply pkg/controllers/dynakube/istio/reconciler.go:240]: okey, it should work.

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 12]
```diff
@@ -157,40 +221,62 @@ func (r *reconciler) reconcileIPServiceEntry(ctx context.Context, ipHosts []dtcl
...
+
+	err := r.client.DeleteServiceEntry(ctx, entryName)
+	if err != nil {
+		return err
+	}
+
+	err = r.client.DeleteVirtualService(ctx, entryName)
+	if err != nil {
+		return err
 	}
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:280]: here if you was not able to delete SE, we never try to delete VS. May be you should use same trick you did with errors.Join?
```
package main

import (
	"errors"
	"fmt"
)

func main() {
	err1 := errors.New("err1")
	var err2 error
	err := errors.Join(err1, err2)
	fmt.Println(err)
	if errors.Is(err, err1) {
		fmt.Println("err is err1")
	}
}
```

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 10 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: fixed in dc69fe66

File: pkg/controllers/dynakube/injection/reconciler.go (hunk 9 above)
Comment (wepudt) [reply pkg/controllers/dynakube/injection/reconciler.go]: fixed in dc69fe66

[the remark this answers is not in the document]
Comment (wepudt) [reply pkg/controllers/dynakube/injection/reconciler.go]: fixed in dc69fe66

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 12 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go:280]: I like it, should improve cleanup.

Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: fixed in c09681fc

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 13]
```diff
@@ -157,37 +221,43 @@ func (r *reconciler) reconcileIPServiceEntry(ctx context.Context, ipHosts []dtcl
...
 
+func (r *reconciler) cleanupIPServiceEntry(ctx context.Context, component string) error {
+	entryName := BuildNameForIPServiceEntry(r.client.Owner.GetName(), component)
+
+	err := r.client.DeleteServiceEntry(ctx, entryName)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:232]: this code can be simplified, you are doing nothing with err. so
```
return r.client.DeleteServiceEntry(ctx, entryName)
```

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 14]
```diff
@@ -54,43 +60,101 @@ func (r *reconciler) ReconcileAPIUrl(ctx context.Context, dynakube *dynatracev1b
...
 	}
 
 	if !dynakube.NeedsActiveGate() {
+		if isIstioConfigured(dynakube, conditionComponent) {
+			log.Info("activegate disabled, cleaning up")
+
+			return r.CleanupIstio(ctx, dynakube, conditionComponent, ActiveGateComponent)
+		} else {
+			return nil
+		}
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:115]:
```suggestion
		if isIstioConfigured(dynakube, conditionComponent) {
			log.Info("activegate disabled, cleaning up")

			return r.CleanupIstio(ctx, dynakube, conditionComponent, ActiveGateComponent)
		}
		return nil
```
else is redundant.

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 15]
```diff
@@ -54,43 +60,101 @@ func (r *reconciler) ReconcileAPIUrl(ctx context.Context, dynakube *dynatracev1b
...
 	err := r.reconcileCommunicationHostsForComponent(ctx, activeGateEndpoints, ActiveGateComponent)
 	if err != nil {
+		setServiceEntryFailedConditionForComponent(dynakube.Conditions(), conditionComponent, err)
+
 		return err
 	}
 
+	if len(activeGateEndpoints) > 0 {
+		setServiceEntryUpdatedConditionForComponent(dynakube.Conditions(), conditionComponent)
+	} else {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:135]: else is redundunt here.

File: pkg/controllers/dynakube/activegate/reconciler.go [hunk 16]
```diff
@@ -107,12 +107,10 @@ func (r *Reconciler) Reconcile(ctx context.Context) error {
...
 
-	if r.dynakube.NeedsActiveGate() {
-		if r.istioReconciler != nil {
-			err = r.istioReconciler.ReconcileActiveGateCommunicationHosts(ctx, r.dynakube)
-			if err != nil {
-				return err
-			}
+	if r.istioReconciler != nil {
+		err = r.istioReconciler.ReconcileActiveGateCommunicationHosts(ctx, r.dynakube)
+		if err != nil {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/activegate/reconciler.go:112]: useless error checking, just return `r.istioReconciler.ReconcileActiveGateCommunicationHosts(ctx, r.dynakube)`

File: pkg/controllers/dynakube/injection/reconciler.go [hunk 17]
```diff
@@ -77,6 +77,15 @@ func (r *reconciler) Reconcile(ctx context.Context) error {
 		return err
 	}
 
+	// do istio reconciliation for CodeModules here to enable cleanup of conditions
+	if r.istioReconciler != nil {
+		err = r.istioReconciler.ReconcileCodeModuleCommunicationHosts(ctx, r.dynakube)
+
+		if err != nil {
+			log.Info("error reconciling istio configuration for codemodules", "error", err)
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/injection/reconciler.go:85]:
```suggestion
			log.Error("error reconciling istio configuration for codemodules", "error", err)
```
may be instead?

File: pkg/controllers/dynakube/istio/conditions.go [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+	}
+	_ = meta.SetStatusCondition(conditions, condition)
+}
+
+func setServiceEntryFailedConditionForComponent(conditions *[]metav1.Condition, component string, err error) {
+	condition := metav1.Condition{
+		Type:    getConditionTypeName(component),
+		Status:  metav1.ConditionFalse,
+		Reason:  "IstioFor" + component + "Failed",
+		Message: "Failed to configure Istio ServiceEntries and VirtualServices for " + component + " with error: " + err.Error(),
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/conditions.go:27]: @0sewa0 do we have limit lenght for message? if err is long ? nvm already found:
```
	// message is a human readable message indicating details about the transition.
	// This may be an empty string.
	// +required
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:MaxLength=32768
	Message string `json:"message" protobuf:"bytes,6,opt,name=message"`
```

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 19]
```diff
@@ -54,43 +60,101 @@ func (r *reconciler) ReconcileAPIUrl(ctx context.Context, dynakube *dynatracev1b
...
 	err := r.reconcileCommunicationHostsForComponent(ctx, oneAgentCommunicationHosts, OneAgentComponent)
 	if err != nil {
+		setServiceEntryFailedConditionForComponent(dynakube.Conditions(), conditionComponent, err)
+
 		return err
 	}
 
+	if len(oneAgentCommunicationHosts) > 0 {
+		setServiceEntryUpdatedConditionForComponent(dynakube.Conditions(), conditionComponent)
+	} else {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:92]: else is redundunt.

> The Go convention is that your "happy path" should be left-aligned and "exceptions" should be indented.

File: pkg/controllers/dynakube/istio/reconciler.go [hunk 20]
```diff
@@ -54,43 +60,101 @@ func (r *reconciler) ReconcileAPIUrl(ctx context.Context, dynakube *dynatracev1b
...
+		meta.RemoveStatusCondition(dynakube.Conditions(), getConditionTypeName(conditionComponent))
+	}
+
 	return nil
 }
 
 func (r *reconciler) ReconcileActiveGateCommunicationHosts(ctx context.Context, dynakube *dynatracev1beta2.DynaKube) error {
+	const conditionComponent = "ActiveGate"
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/istio/reconciler.go:100]: I don't get the point why we need this one, if we have `ActiveGateComponent` from pkg/controllers/dynakube/istio/config.go:16

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 13 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: fixed in a6858713

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 20 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go:100]: If I use `ActiveGateComponent` the condition would then be called `IstioForactivegate` which I don't like. 
Uppercasing the first rune is also no good solution imho- adds to code complexity for only mediocre improvement. Simplest solution imho is to use this const.

Comment (andriisoldatenko) [reply pkg/controllers/dynakube/istio/reconciler.go:100]: but why not to more to all constants? so later we can refactor it. BTW we use ActiveGateComponent only in 1 place. So it makes sense to tweak it and then just apply strings.Lower(ActiveGateComponent) in 2 places.

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 19 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: fixed in c4c2371a

File: pkg/controllers/dynakube/injection/reconciler.go (hunk 17 above)
Comment (wepudt) [reply pkg/controllers/dynakube/injection/reconciler.go]: fixed in c4c2371a

File: pkg/controllers/dynakube/activegate/reconciler.go (hunk 16 above)
Comment (wepudt) [reply pkg/controllers/dynakube/activegate/reconciler.go:112]: fixed in c4c2371a

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 14 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: fixed in c4c2371a

File: pkg/controllers/dynakube/injection/reconciler.go (hunk 17 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/injection/reconciler.go]: this commit hash about different change.

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 19 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/istio/reconciler.go]: wrong commit hash.

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 20 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/istio/reconciler.go:102]: so something like it:
```diff
--- pkg/controllers/dynakube/istio/config.go
+++ pkg/controllers/dynakube/istio/config.go
@@ -13,7 +13,7 @@ var (
 const (
        OperatorComponent   = "operator"
        OneAgentComponent   = "oneagent"
-       ActiveGateComponent = "activegate"
+       ActiveGateComponent = "ActiveGate"
        IstioGVRName        = "networking.istio.io"
        IstioGVRVersion     = "v1beta1"
 )
```


```diff
 func (r *reconciler) ReconcileActiveGateCommunicationHosts(ctx context.Context, dynakube *dynatracev1beta2.DynaKube) error {
-       const conditionComponent = "ActiveGate"
-
        log.Info("reconciling istio components for activegate communication hosts")

        if dynakube == nil {
@@ -106,16 +105,16 @@ func (r *reconciler) ReconcileActiveGateCommunicationHosts(ctx context.Context,
        }

        if !dynakube.NeedsActiveGate() {
-               if isIstioConfigured(dynakube, conditionComponent) {
+               if isIstioConfigured(dynakube, ActiveGateComponent) {
                        log.Info("ActiveGate disabled, cleaning up")

-                       return r.CleanupIstio(ctx, dynakube, conditionComponent, ActiveGateComponent)
+                       return r.CleanupIstio(ctx, dynakube, ActiveGateComponent, strings.ToLower(ActiveGateComponent))
                } else {
                        return nil
                }
        }

-       if !conditions.IsOutdated(r.timeProvider, dynakube, getConditionTypeName(conditionComponent)) {
+       if !conditions.IsOutdated(r.timeProvider, dynakube, getConditionTypeName(ActiveGateComponent)) {
                log.Info("Condition still within time threshold...skipping further reconciliation")

                return nil
@@ -123,17 +122,17 @@ func (r *reconciler) ReconcileActiveGateCommunicationHosts(ctx context.Context,

        activeGateEndpoints := activegate.GetEndpointsAsCommunicationHosts(dynakube)

-       err := r.reconcileCommunicationHostsForComponent(ctx, activeGateEndpoints, ActiveGateComponent)
+       err := r.reconcileCommunicationHostsForComponent(ctx, activeGateEndpoints, strings.ToLower(ActiveGateComponent))
        if err != nil {
-               setServiceEntryFailedConditionForComponent(dynakube.Conditions(), conditionComponent, err)
+               setServiceEntryFailedConditionForComponent(dynakube.Conditions(), ActiveGateComponent, err)

                return err
```

File: pkg/controllers/dynakube/injection/reconciler.go (hunk 17 above)
Comment (wepudt) [reply pkg/controllers/dynakube/injection/reconciler.go]: sorry my bad 76f4aaa0

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 19 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: copy paste error 😞 76f4aaa0

File: pkg/controllers/dynakube/istio/reconciler.go (hunk 20 above)
Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: done in d56e6ba3

Comment (wepudt) [reply pkg/controllers/dynakube/istio/reconciler.go]: done in a9c0115d1d132628574371076b3627808106ae13

File: pkg/controllers/dynakube/activegate/reconciler.go (hunk 16 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/activegate/reconciler.go:112]: Yep, Werner is right. The function continues after this.

**Review by albertogdd — COMMENTED** [review]
Code LGTM, but I deployed it in my cluster with CloudNativeFullStack and Istio and I didn't find the conditions in the Dynakube :thinking: But I think is an issue from my Istio setup

**Review by aorcholski — APPROVED** [review]

**Review by aorcholski — APPROVED** [review]
LGTM. I see new conditions, appropriete condition disappeared after update of dynakube (AG removed)

**Review by 0sewa0 — APPROVED** [review]

**Comment by albertogdd** [comment]
Checked it again this morning and it works :+1:
